### PR TITLE
gluon-web-admin: add option to show/hide password-login and add minimum password length

### PR DESCRIPTION
**Pull request by rubo77**
Or maybe better min 12 characters?

It would be nice, if you get a verbose error message, if the pass is too short, but that needs some more work on gluon-web, i guess.

pw

### CODE REVIEW

**Comment by NeoRaider** [comment]
How about adding a hint about the minimum length to one the texts, for example the label of the text field?

**Comment by Sunz3r** [comment]
pls let the user decide how strong the password should be

**Comment by rubo77** [comment]
If you want a weak password, you can set that ok the console.

If we allow weak passwords​ by default, we'll get complaints, that some users home net was hacked by Freifunk.

We could add a site. Conf option"allow weak passwords"

**Comment by rubo77** [comment]
We could add a site. Conf option"allow weak passwords"

**Comment by rubo77** [comment]
We could add a check for the common mistakes like repeating string, no capital chat, no number,....

**Comment by Sunz3r** [comment]
you are right, hacked by freifunk are very bad.
how about limiting brute force attacks on ssh with iptables?

**Comment by rubo77** [comment]
I added to the text " (min. 8 characters)"

**Comment by rubo77** [comment]
tested and working. I added a screenshot to the first post.

the empty password field is now always red on load, which looks a bit strange.

**Comment by NeoRaider** [comment]
I don't see a reason to allow passwords < 8 characters, that's a very weak limilation.

If you really want a weaker password, you can still set one via SSH.

**Comment by rotanid** [comment]
good idea @rubo77 .
we removed the password option completely in our firmware - is just don't have time to upstream this using a site.conf-option to enable/disable this...

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 1]
```diff
@@ -33,8 +33,9 @@ end
 
 local f_password = Form(translate("Password"),
 	translate(
-                "Alternatively, you can set a password to access you node. Please choose a secure password you don't use anywhere else.<br /><br />"
-                .. "If you set an empty password, login via password will be disabled. This is the default."
+		"Alternatively, you can set a password to access you node. Please choose a "
```
Comment (rotanid) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: although this typo was already there, please fix "you node" -> "your node"

**Comment by rubo77** [comment]
What would you want to control via site conf?

Maybe I find the time

**Comment by rotanid** [comment]
@rubo77 control if the password-ssh-option is in the config mode or not.
we removed it in our gluon fork, but did it by removing the code, as that was easier than making it optional via site.conf
ask me on IRC if something isn't clear

**Comment by rubo77** [comment]
OK, i will add two options to site.conf:

remote = {
  allow_password_login = true, -- optional; default: true
  min_passwordlength = '8'     -- optional; default: 0; set this to 0 to disable a length check
},

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 2]
```diff
@@ -30,80 +31,88 @@ function keys:write(value)
 	end
 end
 
+if ((site.config_mode or {}).remote_login or {}).allow_password_login then
```
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: not sure, how to formulate this, so it defaults to true

**Comment by rubo77** [comment]
it is configurable now in site.conf.

**Comment by rubo77** [comment]
What do you thhink of an option to check the password-strength also?

I could add a codeblock that checks for existance of capital letters, numbers and special chars

File: docs/package/gluon-web-admin.rst [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+site.conf
+---------
+
+config_mode.remote_login.allow_password_login \: optional (defaults to true)
+  If ``allow_password_login`` is set to ``false``, the password section in
+  config mode is hidden
+  
+config_mode.remote_login.min_password_length \: optional (defaults to '8')
+  This stes a minimum limit to the allowed password length. Set this to 0 to
```
Comment (rotanid) [inline docs/package/gluon-web-admin.rst]: "stes" is wrong. also, the sentence is weird... maybe "This sets the minimum allowed password length."

**Comment by rotanid** [comment]
checking the password strength should be optional though, if you add it.
suggestion would be to add some levels like "0 = no check", "1 = at least two of four types of chars" , "2 = at least three of four types" etc. - where types are lowercase letters, uppercase letters, digits, special chars

please also update the pull request title and text and the commit message according to your additional changes

**Comment by NeoRaider** [comment]
Having an option to disable the password form completely make sense in my opinion; the other options just add complexity. We should define simple, sane defaults and keep it at that.

**Comment by rubo77** [comment]
But I want to set the minimum length to 12 characters in Kiel

**Comment by NeoRaider** [comment]
Is this an actual issue? Passwords are only for people who need SSH access. If you know how to use the command line, you should be able to choose a good password...

If you're concerned about stupid users, you really shoud just hide the password form. It will still be possible to set passwords over SSH in config mode (and users who're annoyed by arbitrary password restrictions in the web interface will simply go that route...)

**Comment by rubo77** [comment]
OK, If I remove the length-definition, I will set the minimum Length to 12 characters

**Comment by rubo77** [comment]
OK, I removed the length-definition and set the minimum Length to 12 characters

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 4]
```diff
@@ -30,80 +31,87 @@ function keys:write(value)
 	end
 end
 
+if ((site.config_mode or {}).remote_login or {}).allow_password_login ~= false then
```
Comment (kb-light) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: maybe:
```
if ((site.config_mode or {}).remote_login or {}).allow_password_login == false then
 	-- password login is disabled in site.conf
 	return f_keys
end
[...]
return f_keys, f_password
```
to not indent the whole script

Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: I thought about this earlyer, but then I thought it wouldn't be good readeable to exit the script so early with return. But fine ;=

done

**Comment by rubo77** [comment]
I built it again, and it looks fine.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 5]
```diff
@@ -43,12 +48,15 @@ local s = f_password:section(Section)
...
 local pw1 = s:option(Value, "pw1", translate("Password"))
 pw1.password = true
+pw1.datatype = 'minlength(12)'
 function pw1.cfgvalue()
 	return ''
 end
 
-local pw2 = s:option(Value, "pw2", translate("Confirmation"))
+local pw2 = s:option(Value, "pw2", translate("Confirmation"),
+	translate("Minimum") .. " " .. 12 .. " " .. translate("characters"))
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: Always use `translatef` with format strings instead of splitting up sentences into multiple translation strings (sentences aren't constructed the same way in all languages). The English string should be "%u characters min.".

File: docs/package/gluon-web-admin.rst [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+config_mode.remote_login.allow_password_login \: optional (defaults to true)
+  If ``allow_password_login`` is set to ``false``, the password section in
+  config mode is hidden
+
+Example::
+
+  config_mode = {
+    remote_login = {
+      allow_password_login = true
```
Comment (NeoRaider) [inline docs/package/gluon-web-admin.rst]: Maybe "show_password_form"? Password login is still allowed when the form is disabled.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 7]
```diff
@@ -43,12 +48,15 @@ local s = f_password:section(Section)
 
 local pw1 = s:option(Value, "pw1", translate("Password"))
 pw1.password = true
+pw1.datatype = 'minlength(12)'
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: After thinking about this a bit more, I'm okay with making the limit configurable in site.conf after all.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 8]
```diff
@@ -30,11 +31,15 @@ function keys:write(value)
 	end
 end
 
+if ((site.config_mode or {}).remote_login or {}).allow_password_login == false then
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: site.conf check for the new setting is missing

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 7 above)
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: You'll also need to set *optional* on these fields to allow unsetting the password again. The *write* method must then be adjusted to check for *nil* values.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 8 above)
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: I'll check that if I built again. what exactly do you have in mind here?

**Comment by rubo77** [comment]
I added everything and tested it again. build complete and it works fine:

auswahl_100

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 9]
```diff
@@ -86,6 +94,11 @@ local function set_password(password)
 end
 
 function f_password:write()
+	if pw1.data:len() < min_password_length then
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: Using the datatype is much better than checking manually, as the fields will give direct feedback (background color) then. Nicer textual error messages are a TODO for future improvements in gluon-web.

We're only using the ugly errmessage because we can't express the "same value" constraint using datatypes, ideally, we'd change that to use a similar validation in the future, too.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 10]
```diff
@@ -30,11 +31,17 @@ function keys:write(value)
 	end
 end
 
+local min_password_length = ((site.config_mode or {}).remote_login or {}).min_password_length or 12
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: I'd like to keep the default unchanged from the current situation, so make the default 1.

File: package/gluon-web-admin/i18n/de.po [hunk 11]
```diff
@@ -17,7 +17,7 @@ msgid "Advanced settings"
 msgstr "Erweiterte Einstellungen"
 
 msgid ""
-"Alternatively, you can set a password to access you node. Please choose a "
+"Alternatively, you can set a password to access your node. Please choose a "
```
Comment (NeoRaider) [inline package/gluon-web-admin/i18n/de.po]: This is an unrelated change that affects multiple files. Please split it out into a second commit.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 12]
```diff
@@ -47,7 +54,8 @@ function pw1.cfgvalue()
 	return ''
 end
 
-local pw2 = s:option(Value, "pw2", translate("Confirmation"))
+local pw2 = s:option(Value, "pw2", translate("Confirmation"),
+	translatef("%u characters min.", min_password_length))
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: The length hint should only be shown when min_password_length > 1.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 10 above)
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: But we try to fix an existing serious deficit here, so adding a minimum length is cruicial to security in Freifunk.

And there is no drawback for users, that used a short password before, because there cannot be no check of en existing password that is too short.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 9 above)
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: I had this before, but Then the password field was always red, until you typed something in.

The solution now is much more intuitive and uses the correct warning colors.

File: package/gluon-web-admin/i18n/de.po (hunk 11 above)
Comment (rubo77) [inline package/gluon-web-admin/i18n/de.po]: Sure, this is unrelated, but it only affects files, that we change in this PR anyway. Should I still split it off?

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 9 above)
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: It was red because you didn't define the fields as optional.

File: package/gluon-web-admin/i18n/de.po (hunk 11 above)
Comment (NeoRaider) [inline package/gluon-web-admin/i18n/de.po]: Yes.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 10 above)
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: Usually I would agree, but here we try to fix an existing serious security whole. So adding a minimum length is cruicial for the security of the nodes. I cannot think of a case, where a short password is needed.

Also here there is no drawback for users, that used a short password before, because there cannot be no check of en existing password that is too short.

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 9 above)
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: done

File: package/gluon-web-admin/i18n/de.po (hunk 11 above)
Comment (rubo77) [inline package/gluon-web-admin/i18n/de.po]: done

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua (hunk 12 above)
Comment (rubo77) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: done

**Comment by PowerPan** [comment]
I think every user should have the chance to select his own password.

Maybe it could be better to make a informative message, that the password with the length < $minlength is a weak password but if someone want to use it he should do this. 

It is not our responsibility to enforce security, but we should point it out

**Comment by rubo77** [comment]
As said before: If you want a weak password, you can set it on the console via SSH. And if you want to allow it in your Community in config mode, you can set the min-length to 1 in your `site.conf`.

We are discussing at the moment the default settings if nothing is set in `site.conf`. Those should be set on maximum security, which would be one of the following:

**Comment by rubo77** [comment]
We could set as default **password-login disabled**

**Comment by rubo77** [comment]
We could set a long password as default (10 or 12 chaaracters)

**Comment by rotanid** [comment]
@rubo77 the last time we talked about this in IRC, as far as i remember we said "default disabled, default 12 chars", because we said existing setups aren't touched and everyone can change it in the site.conf if they really want

**Comment by rubo77** [comment]
Ok, I changed the default for `show_password_form` to false

**Comment by rotanid** [comment]
maybe the example in the documentation should reflect the default values, not something random.
also, please update the commit message to reflect the password show/hide

**Comment by rubo77** [comment]
I changed the example, so it shows how to enable it with 12 chars. and changed the text afccordingly

commit message is now "gluon-web-admin: add option to show/hide password-login and add minimum password length"

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 13]
```diff
@@ -30,11 +31,18 @@ function keys:write(value)
 	end
 end
 
+local min_password_length = ((site.config_mode or {}).remote_login or {}).min_password_length or 12
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: You could reduce the code duplication a bit by using a variable for the remote login config:
```lua
local config = (site.config_mode or {}).remote_login or {}
```

File: package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua [hunk 14]
```diff
@@ -43,12 +51,20 @@ local s = f_password:section(Section)
...
 local pw1 = s:option(Value, "pw1", translate("Password"))
 pw1.password = true
+pw1.optional = true
+pw1.datatype = mintype
 function pw1.cfgvalue()
 	return ''
 end
 
-local pw2 = s:option(Value, "pw2", translate("Confirmation"))
+local hint = nil;
```
Comment (NeoRaider) [inline package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua]: Unneeded semicolon and initialization (a variable's value will default to nil when not specified)

**Comment by NeoRaider** [comment]
I'm not sure what happened to the *hint* variable, and the coding style of the latest changes is bad, but I'll just fix it up myself...

**Comment by rubo77** [comment]
the variable hint was just used once, so I thought it would be better to directly add it in place. But ok, I'll keep in mind, that the readeablility is more important than saving a few characters in the code.

**Comment by NeoRaider** [comment]
Merged with minor adjustments.

**Comment by rotanid** [comment]
commit that was merged, for reference:
84d117ff73ac3a2099f851d0c4748cbfd1f41626
